### PR TITLE
Mimirtool: add --output-dir to alertmanager get

### PR DESCRIPTION
**Pull request by edwintye**
#### What this PR does

Seeking some direction from the maintainers and raising the PR for feedback.

We have encountered the same problem where getting the config synced for alert manager using `mimirtool` is quite cumbersome. Looking to fix the issue where the output of `mimirtool alertmanager get` cannot be used for `mimirtool alertmanager load` since the `get` command output a single file that contains both the config and the templates file. Furthermore, the output is not a valid yaml.

To be honest, I am not entirely sure what direction this should go hence this only contains a minimal working enhancement that works on my test setup. For example, 

```shell
# assuming the env variables MIMIR_ADDRESS, MIMIR_TENANT_ID, MIMIR_API_KEY exists and are correct
# build the binary locally, then create the config and template
./mimirtool alertmanager get --log.level="debug" --output-dir="am"
# and assuming that all the templates has file extension .tpl
./mimirtool alertmanager load am/config.yaml am/*.tpl
```

Some of the obvious caveat:
 * Name of the main config is not configurable and fixed as `config.yaml`.
 * When using the flag `--output-dir` the config/templates no longer print to screen.
 * The debug log may be slightly out of place for some people, but since there is only one field `fileName` per line I am not sure `WithFields` is sensible (bit of personal preference here).
 * The flag `--disable-color` is ignored (well, this is fine unless we want color in txt file...).

and there are probably some edge cases which I am not aware of :) .

#### Which issue(s) this PR fixes or relates to

Enhancement for #3241 

#### Checklist

- [x] Tests updated. - Not done because not sure about direction.
- [x] Documentation added. - N/A maybe?
- [ X] `CHANGELOG.md` updated - the order of entries should be `[CHANGE]`, `[FEATURE]`, `[ENHANCEMENT]`, `[BUGFIX]`.
- [ X] [`about-versioning.md`](/docs/sources/mimir/configure/about-versioning.md) updated with experimental features.

### CODE REVIEW

**Comment by leizor** [comment]
The CHANGELOG has just been cut to prepare for the next Mimir release. Please rebase `main` and eventually move the CHANGELOG entry added / updated in this PR to the top of the CHANGELOG document. Thanks!

File: pkg/mimirtool/commands/alerts.go [hunk 1]
```diff
@@ -109,9 +111,33 @@ func (a *AlertmanagerCommand) getConfig(_ *kingpin.ParseContext) error {
...
 
-	return p.PrintAlertmanagerConfig(cfg, templates)
+func (a *AlertmanagerCommand) outputAlertManagerConfigTemplates(config string, templates map[string]string) error {
+	var baseDir string
+	baseDir, err := filepath.Abs(a.OutputDir)
+	if err != nil {
+		return err
+	}
+	log.Debug(fmt.Sprintf("writing the config file to %s", filepath.Join(baseDir, "config.yaml")))
+	err = os.WriteFile(filepath.Join(baseDir, "config.yaml"), []byte(config), os.FileMode(0o666))
```
Comment (dimitarvdimitrov) [inline pkg/mimirtool/commands/alerts.go:128]: nitpick, but can you put `filepath.Join(baseDir, "config.yaml")` in a variable and share it in the two lines? It also avoids repeating the constant

Comment (dimitarvdimitrov) [inline pkg/mimirtool/commands/alerts.go:128]: Let's default to making the file a bit less permissive by default. I'm thinking about running on shared machines and the config can contain API keys and such. Should we go with `0o600`?
```suggestion
	log.Debug(fmt.Sprintf("writing the config file to %s", filepath.Join(baseDir, "config.yaml")))
	err = os.WriteFile(filepath.Join(baseDir, "config.yaml"), []byte(config), os.FileMode(0o600))
```

File: pkg/mimirtool/commands/alerts.go [hunk 2]
```diff
@@ -109,9 +111,33 @@ func (a *AlertmanagerCommand) getConfig(_ *kingpin.ParseContext) error {
...
+	if err != nil {
+		return err
+	}
+
+	for fn, template := range templates {
+		log.Debug(fmt.Sprintf("writing the template file %s", fn))
+		err = os.WriteFile(filepath.Join(baseDir, fn), []byte(template), os.FileMode(0o666))
```
Comment (dimitarvdimitrov) [inline pkg/mimirtool/commands/alerts.go:135]: same comment about a variable for the path. Would also be nice to log the full path in the debug log

File: pkg/mimirtool/commands/alerts.go [hunk 3]
```diff
@@ -109,9 +111,33 @@ func (a *AlertmanagerCommand) getConfig(_ *kingpin.ParseContext) error {
...
+}
 
-	return p.PrintAlertmanagerConfig(cfg, templates)
+func (a *AlertmanagerCommand) outputAlertManagerConfigTemplates(config string, templates map[string]string) error {
+	var baseDir string
+	baseDir, err := filepath.Abs(a.OutputDir)
+	if err != nil {
+		return err
+	}
+	log.Debug(fmt.Sprintf("writing the config file to %s", filepath.Join(baseDir, "config.yaml")))
```
Comment (dimitarvdimitrov) [inline pkg/mimirtool/commands/alerts.go:127]: using Debugf is a bit birefer; same comment applies below
```suggestion
	log.Debugf("writing the config file to %s", filepath.Join(baseDir, "config.yaml"))
```

**Review by dimitarvdimitrov — COMMENTED** [review]
amazing! Thanks a lot for the contribution! I left a few comments, but this is looking great

File: CHANGELOG.md [hunk 4]
```diff
@@ -144,6 +144,7 @@
 ### Mimirtool
 
 * [ENHANCEMENT] Analyze Grafana: Improve support for variables in range. #6657
+* [ENHANCEMENT] Add argument `--output-dir` to `mimirtool alertmanager get` where the config and templates will be written to and can be loaded via `mimirtool alertmanager load` #3241
```
[the remark this answers is not in the document]
Comment (edwintye) [reply CHANGELOG.md]: yes, done.

File: pkg/mimirtool/commands/alerts.go [hunk 5]
```diff
@@ -72,6 +73,7 @@ func (a *AlertmanagerCommand) Register(app *kingpin.Application, envVars EnvVarN
 	// Get Alertmanager Configs Command
 	getAlertsCmd := alertCmd.Command("get", "Get the Alertmanager configuration that is currently in the Grafana Mimir Alertmanager.").Action(a.getConfig)
 	getAlertsCmd.Flag("disable-color", "disable colored output").BoolVar(&a.DisableColor)
+	getAlertsCmd.Flag("output-dir", "The directory where the config and templates will be written to").ExistingDirVar(&a.OutputDir)
```
[the remark this answers is not in the document]
Comment (edwintye) [reply pkg/mimirtool/commands/alerts.go]: yes, I have added extra comment to the argument and also put some info into the docs as suggested.

File: pkg/mimirtool/commands/alerts.go (hunk 1 above)
Comment (edwintye) [reply pkg/mimirtool/commands/alerts.go]: You are right, not even sure why I was thinking of using `0o666` in the first place. Updated.

**Comment by edwintye** [comment]
Thanks a lot for the comments. I have updated as suggested.

File: pkg/mimirtool/commands/alerts.go [hunk 6]
```diff
@@ -109,9 +111,36 @@ func (a *AlertmanagerCommand) getConfig(_ *kingpin.ParseContext) error {
...
+	fileOutputLocation = filepath.Join(baseDir, "config.yaml")
+	log.Debugf("writing the config file to %s", fileOutputLocation)
+	err = os.WriteFile(fileOutputLocation, []byte(config), os.FileMode(0o600))
+	if err != nil {
+		return err
+	}
+
+	for fn, template := range templates {
+		fileOutputLocation = filepath.Join(baseDir, fn)
+		log.Debugf("writing the template file %s", fileOutputLocation)
```
Comment (dimitarvdimitrov) [inline pkg/mimirtool/commands/alerts.go:137]:
```suggestion
		log.Debugf("writing the template file to %s", fileOutputLocation)
```

to match the message above

**Review by dimitarvdimitrov — COMMENTED** [review]
I realized I didn't think about testing either in the last review. I think it will be useful to add a basic happy-path test in this file https://github.com/grafana/mimir/blob/802ce64ac57c15d20fce474cc9fd9505a1e48c48/pkg/mimirtool/commands/alerts_test.go

some pointers
* the flow I'm thinking of is
	1. use the existing test data to load the alertmanager config and templates in memory
	2. write the result to a temporary test directory
	3. load the config and templates from that written out place and assert that they have the same contents as the ones loaded in step i.
* you can use `t.TempDir()` to get a temporary directory to store the files
* there are existing config and template files in `./testdata` next to `alerts_test.go`
* the function `cmd.readAlertManagerConfig()` already loads what's on disk; you can check how `TestReadAlertmanagerConfigTemplates` sets up the template files from the testdata

File: pkg/mimirtool/commands/alerts.go (hunk 6 above)
Comment (edwintye) [reply pkg/mimirtool/commands/alerts.go]: Thanks for your review. I was quite stuck on how I would add a test since there isn't much to use as reference.  I have now added a test as you suggested, and I had to add a minimal valid alert manager config to `./testdata` for `cmd.readAlertManagerConfig()`. An invalid alert manager config will error out (although there isn't a test for that). I am also not entire sure about the naming conventions here on both the (test) function name and the variables, more than happy to change them along with any other suggestions.

Also added `to` to the debug line, not entirely sure how I have missed that initially.

File: CHANGELOG.md [hunk 7]
```diff
@@ -72,6 +72,7 @@
 ### Mimirtool
 
 * [ENHANCEMENT] Analyze Prometheus: set tenant header. #6737
+* [ENHANCEMENT] Add argument `--output-dir` to `mimirtool alertmanager get` where the config and templates will be written to and can be loaded via `mimirtool alertmanager load` #3241
```
[the remark this answers is not in the document]
Comment (edwintye) [reply CHANGELOG.md:75]: ah sorry, I mistakenly thought the hash is to reference the issue rather than the PR, my bad here and will fix.

File: pkg/mimirtool/commands/alerts_test.go [hunk 8]
```diff
@@ -41,3 +42,43 @@ func TestReadAlertmanagerConfigTemplates_collidingNames(t *testing.T) {
...
+		TemplateFiles: []string{
+			filepath.Join(dir, "alertmanager_template.tmpl"),
+		},
+	}
+	cfgResult, templatesResult, err := cmdLoad.readAlertManagerConfig()
+	require.NoError(t, err)
+
+	assert.Equal(t, cfg, cfgResult)
+	assert.Equal(t, len(templates), len(templatesResult))
+	assert.Equal(t, templates, templatesResult)
```
[the remark this answers is not in the document]
Comment (edwintye) [reply pkg/mimirtool/commands/alerts_test.go]: Templates is actually a map rather than a slice, of `map[<fileName>][fileContent]`. However, on second thought this comparison is probably a bit weak for the future so I have updated it and compare element by element.

[the remark this answers is not in the document]
Comment (dimitarvdimitrov) [reply pkg/mimirtool/commands/alerts_test.go]: oh, i thought that `ElementsMatch` works for maps too. Your approach looks good too

**Review by dimitarvdimitrov — APPROVED** [review]
LGTM. Thank you for the contribution and really appreciate the time you spent on this PR!

File: pkg/mimirtool/commands/alerts_test.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (edwintye) [reply pkg/mimirtool/commands/alerts_test.go]: As a first time contributor, thanks a lot for guiding me through this PR!
